### ISSUE TITLE
EntityMetadataNotFoundError with webpack on migrations

### ISSUE THREAD

**Issue by oThams** [typeorm/typeorm#8835]
## Issue Description
When packaging my project with web pack I'm getting an `EntityMetadataNotFoundError` on my migration files. I followed the [FAQ]( https://github.com/typeorm/typeorm/blob/master/docs/faq.md#how-to-use-webpack-for-the-backend) to produce webpack configuration

### Expected Behavior
`findMetadata` successfully identify the metadata

### Actual Behavior
When finding the metadata for entities, the function isn't able to match `metadata.target` with `target`. 
The first test `metadata.target === target` is failing
Then the other tests don't successfully identify the metadata

```
/sample/node_modules/typeorm/data-source/DataSource.js:289
            throw new EntityMetadataNotFoundError_1.EntityMetadataNotFoundError(target);
                  ^

EntityMetadataNotFoundError: No metadata for "Todo" was found.
    at DataSource.getMetadata (/sample/node_modules/typeorm/data-source/DataSource.js:289:19)
    at /sample/node_modules/typeorm/persistence/EntityPersistExecutor.js:69:51
    at Array.forEach (<anonymous>)
    at /sample/node_modules/typeorm/persistence/EntityPersistExecutor.js:62:26
    at Array.map (<anonymous>)
    at EntityPersistExecutor.execute (/sample/node_modules/typeorm/persistence/EntityPersistExecutor.js:59:66)
```

### Steps to Reproduce

1. create a folder
2. unzip [project.zip](https://github.com/typeorm/typeorm/files/8392251/project.zip) in that folder
3. install packages `npm install` 
4. build with webpack `npm run build`
5. run mini project `npm run start`

### My Environment

| Dependency          | Version  |
| ---                 | ---      |
| Operating System    |   macOS 12.3       |
| Node.js version     | 16.13.0  | 
| Typescript version  | 4.6.3  | 
| TypeORM version     | 0.3.4  | 

### Relevant Database Driver(s)

| DB Type           | Reproducible  |
|-------------------| ---           |
| `aurora-mysql`    | no            |
| `aurora-postgres` | no            |
| `better-sqlite3`  | no            |
| `cockroachdb`     | no            |
| `cordova`         | no            |
| `expo`            | no            |
| `mongodb`         | no            |
| `mysql`           | no            |
| `nativescript`    | no            |
| `oracle`          | no            |
| `postgres`        | no            |
| `react-native`    | no            |
| `sap`             | no            |
| `sqlite`          | yes            |
| `sqlite-abstract` | no            |
| `sqljs`           | no            |
| `sqlserver`       | no            |

### Are you willing to resolve this issue by submitting a Pull Request?

 - ✖️ Yes, I have the time, and I know how to start.
 - ✅ Yes, I have the time, but I don't know how to start. I would need guidance.
 - ✖️ No, I don’t have the time, but I can support (using donations) development.
 - ✖️ No, I don’t have the time and I’m okay to wait for the community / maintainers to resolve this issue.

**Comment by codeninja** [comment]
I was noticing this behavior as well and traced it down to paths. 

Instead of this:
```
import { EntityOne, EntityTwo } from "../../"
```

I needed to specify the path to each: 
```
import { EntityOne } from '../../entities/entityFile'
import { EntityTwo } from '../../entities/entityFile'
```

We've made a new rule to use explicit paths for every import. Much more verbose but it seemed to resolve our particular issue.

**Comment by wcldyx** [comment]
I have the same problem.
There is no problem in the development environment.But this problem can also occur after packaging in production

**Comment by wcldyx** [comment]
I have the same problem.
There is no problem in the development environment.But this problem can also occur after packaging in production

**Comment by alexandregrm** [comment]
Hello,

I had the same problem using webpack and migration with seeds. 

## TL;DR

### For quick fix : 
Use entities as string and not object for your migration seed files
```
queryRunner.manager.create<Todo>("Todo", todo); <=== ✅ using string entity
queryRunner.manager.create<Todo>(Todo, todo); <===  ❌ using object entity
const todo = new Todo(data); <== ❌Using class constructor
```
### Suggestion of dev : see below

## What I understood of this problem

### 1st problem : tells the datasource where entities are located

Depending on your bundle configuration you can :
- Transpile / build each entity and using something like ```entities: [`${rootDir}/../entities/*{.ts,.js}\`],``` in the typeorm config
- Build the datasource with your entities embedded ( I followed [this article](https://stackoverflow.com/questions/66169705/bundle-a-nestjs-typeorm-application-with-webpack))
  - I changed the typeorm config from dynamic import `entities.js` to static `entities: [Todo, Table, ...]` (easier with my configuration )
  - bundle the datasource
- Use the require.context stuff with webpack ( I didn't try hard because I had issues with jest, but perhaps it's possible)
- Do nothing as entities are already in JS

### 2nd problem : tells the datasource that entities are same

When you run the migration there is a conflict between the target from the datasource and the target from the builded migration (webpack build everything needed for the migration in one file). 

So when you tried to compare "`metadata.target === target`" it's throwing an error because objects are different. For example, one came from the "init.migration.js" file and the second one from "entities/table.js".

This error is thrown from `datasource::findMetadata` function because there is no match between entity asked and entities in datasource. Using target entities with name solves the problem.\
If we look at the code we can see the following lines :
```
if (metadata.target === target) <=== ❌ fail
                return true;
...
if (typeof target === "string") {
    ...
    else {
        return (metadata.name === target || <== ✅ use this
            metadata.tableName === target);
    }
    ...
}
```

In your case you can remplace 
```
const todo = new Todo();
todo.description="TODO";
todo.done=false;
queryRunner.manager.save(todo);
```
with 
```
const todo = queryRunner.manager.create<Todo>("Todo", {description: "TODO", done: false});
queryRunner.manager.save(todo);
```

Using Todo as class will fail as object are different for JS.

**DEV suggestion for datasource :**

I can suggest typeorm to add another matching type for the `datasource::findMetadata` as webpack build the class as function constructor.

Maybe something like this : 
```
if (typeof target === 'function') {
  const targetName = (new target()).constructor.name; // Better solution ?
  return metadata.name === targetName ||
    metadata.tableName === targetName;
}
```
Changing this in the datasource code works in my configuration but I'm not sure about side effect with all features of typeorm.

I hope this will help for this problem or another similar :)

**Comment by AhsanRiyad** [comment]
any update?
